airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py

```diff
@@ -86,6 +86,8 @@ def set_initial_state(self, stream_state: StreamState) -> None:
         for state in stream_state["states"]:
             self._cursor_per_partition[self._to_partition_key(state["partition"])] = self._create_cursor(state["cursor"])
 
+        self._partition_router.set_parent_state(stream_state)
```

Comment:
this deserves a comment as it's not obvious why we need to update the router's parent state

Reply:
Added comment

Reply:
Stream slicers and partition routers don't keep track of the cursor and don't have a `set_initial_state` method. `set_parent_state` sets the state to the parent streams that partition routers operate on, if such parent streams exist.

Reply:
seems to me this means we want to add a `set_initial_state` method to the routers - from the caller's perspective, it doesn't really matter that it's setting a parent state. It's an implementation detail. 

Furthermore, not all stream slicers know or care about parents (eg list partition router)

Reply:
edit: I do think the state being set is somewhat different, but the interface feels very strange to me for two reasons
1. Not every stream slicers should know or care about parent streams
2. even from the caller, the fact that it's updating a parent stream state is irrelevant. [It's just passing the whole state object](https://github.com/airbytehq/airbyte/pull/38211/files#diff-cb24ede0b6b09974da9a16d25d8fa5f772ae497c742224f08dac1bdc1a903f9dR119)

Reply:
The same goes with `set_initial_state` for stream slicers, as not every stream slicer cares about the state. Introducing a `set_initial_state` method could create confusion, implying that partition routers are somehow dependent on the state. `set_parent_state` clearly articulates that it is only connected to the parent streams, as the stream interacts with the parent stream through the partition router.

Reply:
yeah, that's exactly the point - we're now saying that `SubstreamPartitionRouter` and `CartesianProductStreamSlicer` are incremental and have a concept of state they need to keep track of.

I'm not sure if it'll make the code simpler or more complicated, but conceptually, this makes me think the `SubstreamPartitionRouter` should have a `Cursor` when `incremental_dependency` is True.

This would essentially mean that instead of keeping the parent state in a dict, it could pass the parent records to its cursor's observe method while iterating over them. 

I feel 4/10 strongly on this issue so not worth blocking on.

Reply:
I changed the methods' names. Regarding using a Cursor to keep track of the state, it is already managed by the cursor of parent streams. Partition routers just retrieve the state from the internal parent streams.